File: src/org/elixir_lang/mix/importWizard/ImportedOtpApp.java
package org.elixir_lang.mix.importWizard;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.util.containers.ContainerUtil;
import org.elixir_lang.mix.util.ElixirScriptFileUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by zyuyou on 15/7/2.
 */
public class ImportedOtpApp {
  private final String myName;
  private final VirtualFile myRoot;
  private final Set<String> myDeps = ContainerUtil.newHashSet();
  private VirtualFile myIdeaModuleFile;
  private Module myModule;

  public ImportedOtpApp(@NotNull VirtualFile root, @NotNull final VirtualFile appMixFile){
    // todo: should use psi-pattern to get myName.
    String appName = "";
    try{
      String firstLine = ElixirScriptFileUtil.readLine(appMixFile.getInputStream());
      Pattern pattern = Pattern.compile("defmodule ([A-Z]{1}\\w*).Mixfile do");
      Matcher matcher = pattern.matcher(firstLine);
      appName = matcher.matches() ? matcher.group(1).toLowerCase() : "";
    }catch (Exception ignored){
    }

    myName = appName;
    myRoot = root;

    ApplicationManager.getApplication().runReadAction(new Runnable() {
      @Override
      public void run() {
        addInfoFromAppMixFile(appMixFile);
      }
    });
  }

  @NotNull
  public String getName(){
    return myName;
  }

  @NotNull
  public VirtualFile getRoot(){
    return myRoot;
  }

  @NotNull
  public Set<String> getDeps(){
    return myDeps;
  }

  public void setIdeaModuleFile(@Nullable VirtualFile ideaModuleFile){
    myIdeaModuleFile = ideaModuleFile;
  }

  @Nullable
  public VirtualFile getIdeaModuleFile(){
    return myIdeaModuleFile;
  }

  public Module getModule(){
    return myModule;
  }

  public void setModule(Module module){
    myModule = module;
  }

  @Override
  public String toString() {
    return myName + " (" + myRoot + ")" ;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (obj == null || getClass() != obj.getClass()) return false;

    ImportedOtpApp that = (ImportedOtpApp) obj;

    return myName.equals(that.myName) && myRoot.equals(that.myRoot);
  }

  @Override
  public int hashCode() {
    int result = myName.hashCode();
    result = 31 * result + myRoot.hashCode();
    return result;
  }

  private void addInfoFromAppMixFile(@NotNull VirtualFile appMixFile){
    // todo: get deps
  }

}

File: src/org/elixir_lang/mix/importWizard/ImportedOtpApp.kt
package org.elixir_lang.mix.importWizard

import com.ericsson.otp.erlang.OtpErlangAtom
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.util.Computable
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiFileFactory
import org.elixir_lang.ElixirFileType
import org.elixir_lang.psi.ElixirAccessExpression
import org.elixir_lang.psi.ElixirFile
import org.elixir_lang.psi.ElixirList
import org.elixir_lang.psi.impl.call.macroChildCallList
import org.elixir_lang.psi.impl.keywordValue
import org.elixir_lang.structure_view.element.CallDefinitionClause.isPublicFunction
import org.elixir_lang.structure_view.element.CallDefinitionClause.nameArityRange
import java.io.IOException
import java.nio.file.Paths

private fun app(appMixFile: VirtualFile): String =
    try {
        app(appMixFile, text(appMixFile))
    } catch (e: IOException) {
        appFromPath(appMixFile)
    }

private fun app(appMixFile: VirtualFile, appMixFileText: String): String {
    val elixirFile = elixirFile(appMixFileText)
    val appList = appList(elixirFile)

    return if (appList.isEmpty()) {
        appFromPath(appMixFile)
    } else {
        appList.first()
    }
}

private fun appList(elixirFile: ElixirFile): List<String> {
    val modulars =  elixirFile
            .modulars()

    return modulars.asSequence()
            .flatMap { modular ->
                modular.macroChildCallList().asSequence()
            }
            .filter { call ->
                isPublicFunction(call) && nameArityRange(call)?.let {
                    it.first == "project" && it.second?.containsInteger(0) == true
                } == true
            }
            .flatMap { projectCallDefinition ->
                projectCallDefinition.doBlock?.stab?.stabBody?.children?.asSequence() ?: emptySequence()
            }
            .filterIsInstance<ElixirAccessExpression>()
            .mapNotNull(ElixirAccessExpression::getList)
            .mapNotNull(ElixirList::getKeywords)
            .mapNotNull { keywordList ->
                keywordList.keywordValue("app")
            }
            .map { keywordValue ->
                keywordValue.quote()
            }
            .filterIsInstance<OtpErlangAtom>()
            .map(OtpErlangAtom::atomValue)
            .toList()
}

private fun appFromPath(appMixFile: VirtualFile): String = Paths.get(appMixFile.path).parent.fileName.toString()

fun <T> computeReadAction(computable: Computable<T>): T =
        ApplicationManager.getApplication().runReadAction(computable)

private fun elixirFile(text: String): ElixirFile = computeReadAction(Computable {
    val defaultProject = ProjectManager.getInstance().defaultProject

    PsiFileFactory
            .getInstance(defaultProject)
            .createFileFromText("mix.exs", ElixirFileType.ElixirScriptFileType.INSTANCE, text) as ElixirFile
})

private fun text(virtualFile: VirtualFile): String = computeReadAction(Computable {
    VfsUtil.loadText(virtualFile)
})

class ImportedOtpApp(val root: VirtualFile, appMixFile: VirtualFile) {
    val deps = mutableSetOf<String>()
    var ideaModuleFile: VirtualFile? = null
    var module: Module? = null
    val name = app(appMixFile)

    override fun toString(): String = "$name ($root)"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val that = other as ImportedOtpApp

        return name == that.name && root == that.root
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + root.hashCode()
        return result
    }
}
